src/wallet/wallet.cpp

```diff
@@ -426,27 +442,243 @@ ZcashdUnifiedSpendingKey CWallet::GenerateNewUnifiedSpendingKey() {
                         "CWallet::GenerateNewUnifiedSpendingKey(): Writing HD chain model failed");
             }
 
-            return usk.value();
+            return std::make_pair(usk.value(), accountId);
         }
     }
 }
 
-std::optional<libzcash::ZcashdUnifiedSpendingKey> CWallet::GenerateUnifiedSpendingKeyForAccount(libzcash::AccountId accountId) {
+std::optional<libzcash::ZcashdUnifiedSpendingKey>
+        CWallet::GenerateUnifiedSpendingKeyForAccount(libzcash::AccountId accountId) {
+    AssertLockHeld(cs_wallet); // mapUnifiedAccountKeys
+
     auto seed = GetMnemonicSeed();
     if (!seed.has_value()) {
         throw std::runtime_error(std::string(__func__) + ": Wallet has no mnemonic HD seed.");
     }
 
     auto usk = ZcashdUnifiedSpendingKey::ForAccount(seed.value(), BIP44CoinType(), accountId);
     if (usk.has_value()) {
```

Comment:
This is making changes to the wallet db that are not transactional but probably should be. Please file a ticket to fix.

Reply:
https://github.com/zcash/zcash/issues/5444